### ISSUE TITLE
Managing state when refactoring grains?

### ISSUE THREAD

**Issue by BenjaminGibbs** [dotnet/orleans#61]
As projects move forward requirements change and code ends up being refactored, this could as simple as a namespace being changed, or Oect1Grain being renamed to Object2Gain etc..

What is the recommend way of handling state when this happens? as my understanding from looking at how the state is stored, the grain type is linked to the state record (for example in Azure Storage it becomes the RowKey). Which would imply that refactoring/renaming the grain or namespace would break the reading of the previous state when the service is updated.

One possible solution could to be to add an attribute to the grain (like we do with the storage provider) which allows us to specify the Type as string to override the default type that would be provided enabling backwards compatibility with old state? 

I'd would be interested to hear how others are approaching the scenario, or point out something I may have missed.

**Comment by jkonecki** [comment]
That very much depends on the provider being used but I would agree that
there should be an extension point for the storage provider to override a
default key.

I solved that problem in other Event Sourced projects by defining an
interface that a grain could implement and extending the storage provider
or check if the grain implements this interface before
serialisation/deserialisation.

I could create a PR if this approach is accepted.

**Comment by sergeybykov** [comment]
We would only encourage to critically rethink the persistence provider model. It is currently very basic, and doesn't support some obvious cases like this. Adding an attribute to override the default grain class to row key or even to table name mapping may be a good option.

Jakub went even further by adding an interface to give grain class a way to override serialization/deserialization of its state. My personal feeling that some of these approached may be complementary, to support both basic and more advanced scenarios. One things that concerns me is potential coupling of grain classes to specific providers.

**Comment by jkonecki** [comment]
It's more a coupling of the *state* and provider not the grain itself.

I think that if we want to keep the grain/state totally agnostic of the persistence than the provider framework must allow enough configuration to support cases like migration.

In the example from @BenjaminGibbs the change of the namespace affects the provider's convention of key generation. So either the provider allows the convention to be reconfigured or the key generation is extracted out from the provider to a separate service which providers will depend on.

**Comment by BenjaminGibbs** [comment]
I would agree with avoiding coupling the grain class to the specific provider.

Having had a bit more of a deep dive in to the codebase (I have only been using it a few days), the *GrainState* class passes *grainType (string)* in to the write/read/clear state, so being able to just replace that parameter for something else would solve the problem on a basic level, after all this represents the 'type' of grain (as in domain type not class type) it should make migration painless and shouldn't be too hard to implement, after all a 'user' is always going to be a 'user' no matter what namespace, or class/grain name.

I am still very new to Orleans, but it appears if the *StorageProviderAttribute* was extended to accept an additional optional string parameter then if *void SetupStorageProvider* (Catalog.cs Ln 565) was adjusted to set *grainTypeName* to the new parameter if it is specified, if not revert to back the current format. (to maintain backwards compatibility)

*SetupStorageProvider* already reads the *StorageProviderAttribute*, and at the same time is responsible for setting *grainTypeName* which ends up as the string that is passed in to the write/read/clear state subroutines which is causing the problems when refactoring.

It would be a small change that should be pretty quick, it would maintain compatibility with code already in production, first by not breaking their current code, but also if they do end up refactoring any part of their code they could set the new parameter as the current *grainTypeName*.

I am just conscious that once people start putting Orleans into production, a refactor will happen and then assuming they are using the state as designed then the data becomes inaccessible.

**Comment by sergeybykov** [comment]
@jthelin-microsoft - Any thoughts?

**Comment by yevhen** [comment]
Regarding the subject :

1. I don't think that it's good idea for an actor framework to also include persistence, from SRP point-of-view.
2. Generic persistence cannot satisfy all use-cases. It's a lowest common denominator. Declarative persistence is a fallacy.
3. The current design is broken due to mixing different concerns: `GrainState` is both an in-memory holder of data and unit-of-work (imperative `Write\Clear`). Which is confusing at the best.
4. Also, it requires to create completely unnecessary state interface declarations, which simply breaks encapsulation due to public exposure of actor internals. State is an internal detail.

To be honest, I think persistence providers in Orleans is just a workaround for not having proper dependency injection. Otherwise, you would be able to inject you own persistence abstractions directly into the grain, ie `Repositories`, `Connections`, whatever.

Think about it: the only thing which Orleans does is creating an instance of provider and automatically calling provider's `ReadStateAsync()` during grain `Activate()`. You still need to implement state check-pointing, anyway (`Write\Clear`). So what's the added benefit? And with all those constraints, limitations and abuse. I have no idea.

Even automatic reading of state is inconvenient for some use-cases, where I want to control _when_ state is _actually_ read from the storage. The inconvenience of generic persistence is nicely illustrated in @jkonecki EvenSourcing provider. ES just doesn't fit into Orleans' persistence provider model.

Generic persistence in Orleans, only brings confusion and troubles with no added benefit and it will be even more problems with it in the future (once people start actively using it). Don't repeat Akka's "let's bake everything and the kitchen sink into the framework". 

To summarize: Less is more. Try to convince me otherwise :)

**Comment by ReubenBond** [comment]
It's incorrect to say that Orleans' persistence provides "no added benefit".

Persistence in Orleans is not slated as a panacea, it's simply a way to get running quickly. It satisfies that goal. It's not as though Orleans pushes persistence onto users.

We refer to Orleans as an Actor framework, but it's more than that: it's a framework for easily writing fault-tolerant, distributed systems. There are several problems in that domain. Persistence is one of them.

Akka also includes an optional persistence module (built around event sourcing). Perhaps you would prefer if Orleans also made persistence an optional dependency?

**Comment by jkonecki** [comment]
@yevhen I agree I had to jump through some hoops to get ES provider in. It would be benefitial in my opinion to have more flexibility wrt persistance like you suggest.

I really like the fact that grains are separated from their states. What I believe we need is having full control how and when that state is read / written. 

WRT dependency injection I think we should look at the unified approach that is being made in ASP.NET 5 and probably piggyback on it.

**Comment by sergeybykov** [comment]
I agree with @ReubenBond - Orleans is a framework for building distributed apps. It just so happened that the actor model was a good fit for the task. Persistence is something that pretty much any app has to deal with, and most end up adding a storage library/layer for that. Orleans persistence was meant to be an opt-in 'easy' option for connecting app's storage layer to the activation logic.

I agree with @jkonecki - We need a more robust interface to get the app full control of how and when the state is read/written. I think this can be done in addition to today's simple model, as an optional extension, by maybe I'm delusional.

I see three categories of app cases with regards to persistence:

1. The simple option working as is. We are not quite there yet functionality wise. Need better declarative mapping of types to tables, support for simple migrations, etc.

2. More advanced persistence interface, hopefully, a superset of 1. Direct control over serialization, partial on-demand read/write of fields, support for event sourcing, etc. To me, @jkonecki started on that path.

3. App logic handles persistence on its own, hooks up to the activation process via OnActivateAsync. Nothing for us to do.

I see dependency injection being orthogonal to persistence. Not sure why they should be bundled together.

**Comment by gabikliot** [comment]
I like how @sergeybykov ordered the options. From the more automatically managed by Orleans to the least managed.
There is also potentially a fourth option (zero option in this ordering), even more managed by Orleans.
It will be like now and in addition we checkpoint (do WriteASync) after every operation. That way the app code does not need to explicitly checkpoint at all. In fact, we went back and forth a lot between that option and what we finally picked (app does explicit checkpoints).
In that new option we could also make it even more managed, for example by automatically rolling back the state in case of failures. There are a lot of possibilities here.

And it totally fine I think that one option does not cover all cases.

**Comment by vyshkant** [comment]
@sergeybykov after 4 years is there any working migration flow? If there is not, do we have the issues we need to solve to get it working?

**Comment by sergeybykov** [comment]
Still no magic. I see many people putting state migration logic into their custom storage provider code. @ReubenBond has a side project of version tolerant serializer, but it's unclear when it could be leveraged for state migration.

**Comment by ReubenBond** [comment]
This can be achieved using JSON or a custom storage provider or by interfacing with the database outside of Orleans' simplistic persistence model.

The abovementioned version tolerant serializer currently lives on GitHub here: https://github.com/ReubenBond/Hagar and is being published to NuGet.

See also #2653 and a potential solution in #6936.
